File: lib/java/opentoken-cli/src/main/java/com/truveta/opentoken/cli/processor/TokenEncryptionProcessor.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.cli.processor;

import java.io.IOException;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.cli.io.TokenReader;
import com.truveta.opentoken.cli.io.TokenWriter;
import com.truveta.opentoken.tokens.Token;
import com.truveta.opentoken.tokentransformer.EncryptTokenTransformer;

/**
 * Process hashed tokens for encryption.
 * <p>
 * This class is used to read hashed tokens from input source,
 * encrypt them, and write the encrypted tokens to the output data source.
 */
public final class TokenEncryptionProcessor {

    private static final Logger logger = LoggerFactory.getLogger(TokenEncryptionProcessor.class);

    private TokenEncryptionProcessor() {
    }

    /**
     * Reads hashed tokens from the input data source, encrypts them, and
     * writes the result back to the output data source.
     * 
     * @param reader    TokenReader providing hashed token rows
     * @param writer    TokenWriter for output
     * @param encryptor The encryption transformer
     * @throws IOException if an I/O error occurs
     */
    public static void process(TokenReader reader,
                                TokenWriter writer,
                                EncryptTokenTransformer encryptor) throws IOException {
        long rowCounter = 0;
        long encryptedCounter = 0;
        long errorCounter = 0;

        while (reader.hasNext()) {
            Map<String, String> row = reader.next();
            rowCounter++;
            
            String token = row.get(TokenConstants.TOKEN);
            
            // Encrypt the token if it's not blank
            if (token != null && !token.isEmpty() && !Token.BLANK.equals(token)) {
                try {
                    String encryptedToken = encryptor.transform(token);
                    row.put(TokenConstants.TOKEN, encryptedToken);
                    encryptedCounter++;
                } catch (Exception e) {
                    logger.error("Failed to encrypt token for RecordId {}, RuleId {}: {}", 
                               row.get(TokenConstants.RECORD_ID), row.get(TokenConstants.RULE_ID), e.getMessage());
                    errorCounter++;
                    // Keep the hashed token in case of error
                }
            }
            
            // Write token
            writer.writeToken(row);

            if (rowCounter % 10000 == 0) {
                logger.info(String.format("Processed \"%,d\" tokens", rowCounter));
            }
        }

        logger.info(String.format("Processed a total of %,d tokens", rowCounter));
        logger.info(String.format("Successfully encrypted %,d tokens", encryptedCounter));
        if (errorCounter > 0) {
            logger.warn(String.format("Failed to encrypt %,d tokens", errorCounter));
        }
    }
}

File: lib/python/opentoken-cli/src/main/opentoken_cli/processor/token_encryption_processor.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import logging

from opentoken.tokens.token import Token
from opentoken.tokentransformer.encrypt_token_transformer import EncryptTokenTransformer
from opentoken_cli.io.token_reader import TokenReader
from opentoken_cli.io.token_writer import TokenWriter
from opentoken_cli.processor.token_constants import TokenConstants

logger = logging.getLogger(__name__)


class TokenEncryptionProcessor:
    """
    Process hashed tokens for encryption.

    This class is used to read hashed tokens from input source,
    encrypt them, and write the encrypted tokens to the output data source.
    """

    @staticmethod
    def process(reader: TokenReader, writer: TokenWriter, encryptor: EncryptTokenTransformer):
        """
        Reads hashed tokens from the input data source, encrypts them, and
        writes the result back to the output data source.

        Args:
            reader: Iterator providing hashed token rows
            writer: Writer instance with write_token method
            encryptor: The encryption transformer
        """
        row_counter = 0
        encrypted_counter = 0
        error_counter = 0

        for row in reader:
            row_counter += 1

            token = row.get(TokenConstants.TOKEN, "")

            # Encrypt the token if it's not blank
            if token and token != Token.BLANK:
                try:
                    encrypted_token = encryptor.transform(token)
                    row[TokenConstants.TOKEN] = encrypted_token
                    encrypted_counter += 1
                except Exception as e:
                    logger.error(
                        f"Failed to encrypt token for RecordId {row.get(TokenConstants.RECORD_ID)}, "
                        f"RuleId {row.get(TokenConstants.RULE_ID)}: {e}"
                    )
                    error_counter += 1
                    # Keep the hashed token in case of error

            writer.write_token(row)

            if row_counter % 10000 == 0:
                logger.info(f'Processed "{row_counter:,}" tokens')

        logger.info(f"Processed a total of {row_counter:,} tokens")
        logger.info(f"Successfully encrypted {encrypted_counter:,} tokens")
        if error_counter > 0:
            logger.warning(f"Failed to encrypt {error_counter:,} tokens")
